cloudigrade/account/reports/__init__.py

```diff
@@ -0,0 +1,179 @@
+
+    Returns:
+        list(InstanceEvent): All events relevant to the report parameters.
+    """
+    # Get the nearest event before the reporting period for each instance.
+    account_filter = cloud_helper.instance_account_filter()
+    instances_before = Instance.objects.filter(
+        account_filter & models.Q(instanceevent__occurred_at__lt=start)
+    ).annotate(occurred_at=models.Max('instanceevent__occurred_at'))
+
+    # Build a filter list for the events found before the period.
+    event_filters = [
+        models.Q(instance__id=i.id, occurred_at=i.occurred_at)
+        for i in list(instances_before)
+    ]
+
+    # Add a filter for the events *during* the reporting period.
+    event_filters.append(
+        cloud_helper.event_account_filter() &
+        models.Q(occurred_at__gte=start, occurred_at__lt=end)
+    )
+
+    # Reduce all the filters with the "or" operator.
+    event_filter = functools.reduce(operator.ior, event_filters)
+    events = InstanceEvent.objects.filter(event_filter).select_related()\
+        .order_by('instance__id')
+    return events
+
+
+def _group_related_events(events, cloud_helper):
+    """
+    Group the events by multiple dimensions in nested dicts.
+
+    At the outer later, the keys are "product type" strings. These are a
+    combination of (virtual) hardware definition and found software product
+    that ultimately each needs to be reported with a sum total of running time.
+    At the next layer, the keys are the instance IDs. Finally, each of those
+    instance ID keys points to a list of InstanceEvents that belong to it.
+
+    This "double nesting" of product+instance is necessary because instance IDs
+    alone would not be sufficient for grouping because some cloud providers
+    allow resizing of an existing instance without changing/replacing its ID.
+
+    For example, a returned structure might look like:
+
+        {
+            'aws-t2.micro-rhel7': {
+                'i-05f78714782d7979b': [
+                    <InstanceEvent object>
+                ]
+            },
+            'aws-t2.nano-rhel7': {
+                'i-05f64764381d6970a': [
+                    <InstanceEvent object>
+                ],
+                'i-05f78714782d7979b': [
+                    <InstanceEvent object>,
+                    <InstanceEvent object>,
+                    <InstanceEvent object>
+                ]
+            }
+        }
+
+    Args:
+        events (list(InstanceEvent)): Events to group
+        cloud_helper (helper.ReportHelper): Helper for cloud-specific things
+
+    Returns:
+        dict: Results as a dict-of-dicts-of-lists as described above.
+    """
+    grouped_events = collections.defaultdict(
+        functools.partial(collections.defaultdict, list)
+    )
+
+    for event in events:
+        product_key = cloud_helper.get_event_product_identifier(event)
+        instance_key = cloud_helper.get_event_instance_identifier(event)
```

Comment:
Does the Django ORM populate the instance object when pulling the initial events or will this make another DB call for each event?